bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/BulkMutation.java

```diff
@@ -108,78 +151,171 @@ MutateRowsRequest toRequest(){
      * {@link MutateRowsResponse} is complete.
      */
     void addCallback(ListenableFuture<MutateRowsResponse> bulkFuture) {
-      FutureCallback<MutateRowsResponse> callback = new FutureCallback<MutateRowsResponse>() {
-        @Override
-        public void onSuccess(MutateRowsResponse result) {
-          Iterator<Status> statuses = result.getStatusesList().iterator();
-          Iterator<SettableFuture<Empty>> entries = futures.iterator();
-          while (entries.hasNext() && statuses.hasNext()) {
-            SettableFuture<Empty> future = entries.next();
+      Futures.addCallback(
+          bulkFuture,
+          new FutureCallback<MutateRowsResponse>() {
+            @Override
+            public void onSuccess(MutateRowsResponse result) {
+              handleResult(result);
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              setFailure(t);
+            }
+          });
+    }
+
+
+    private synchronized void handleResult(MutateRowsResponse result) {
```

Comment:
This method is on the long side. Any clean way to factor some parts out to smaller methods?